### PR TITLE
Implement `index clear` and `source reset-checkpoint` CLI commands

### PR DESCRIPTION
**Pull request by guilload**
### Description
Per title.

### How was this PR tested?
- added unit test
- ran `make test-all` command

### CODE REVIEW

**Comment by guilload** [comment]
As discussed offline with @fulmicoton, we currently don't have a pressing need for this API. We will revisit when necessary.

**Comment by guilload** [comment]
Reopened. See #1631.

**Comment by guilload** [comment]
This PR is ready for review.

I have introduced two new metastore APIs, which completely clears the checkpoints (keys are removed)
- `reset_index_checkpoint(index_id)`
- `reset_source_checkpoint(index_id, source_id)`

I have introduced two new CLI commands:
- `index reset-checkpoint --index <index ID>`
- `source reset-checkpoint --index <index ID> --source <source ID>`

**Comment by guilload** [comment]
PTAL!

**Comment by kstaken** [comment]
FYI, in relation to this as part of the Kafka offset storage I've been working on metastore level changes to move checkpoints into a separate table on Postgres backed metastores. As part of that I'm adding two APIs, one of which would overlap with this.

```
    /// Retrieves the current checkpoints for an index
    ///
    /// Returns a `IndexCheckpoint` for the provided `index_id` and optionally filtered by
    /// `source_id`, `resource` and `group_id`.
    /// If no checkpoint has been stored yet the result will be an empty `IndexCheckpoint`.
    async fn index_checkpoint(
        &self,
        index_id: &str,
        source_id: Option<String>,
        resource: Option<String>,
        group_id: Option<String>,
    ) -> MetastoreResult<IndexCheckpoint>;

    /// Removes the current checkpoints for an index
    ///
    /// Removes all checkpoints that match the provided `index_id` and optionally filtered by
    /// `source_id`, `resource` and `group_id`.
    /// If no checkpoint has been stored it will not be considered an error.
    async fn delete_index_checkpoint(
        &self,
        index_id: &str,
        source_id: Option<String>,
        resource: Option<String>,
        group_id: Option<String>,
    ) -> MetastoreResult<()>;
```

File: quickwit-cli/src/source.rs [hunk 1]
```diff
@@ -65,6 +65,14 @@ pub fn build_source_command<'a>() -> Command<'a> {
                     arg!(--index <INDEX_ID> "ID of the target index"),
                 ])
             )
+        .subcommand(
+            Command::new("reset-checkpoint")
+                .about("Clears a source checkpoint. This operation is destructive and cannot be undone. Proceed with caution.")
```
Comment (fmassot) [inline quickwit-cli/src/source.rs:70]:
```suggestion
                .about("Resets a source checkpoint. This operation is destructive and cannot be undone. Proceed with caution.")
```
```suggestion
                .about("Resets a source checkpoint. This operation is destructive and cannot be undone. Proceed with caution.")
```

File: quickwit-metastore/src/checkpoint.rs [hunk 2]
```diff
@@ -161,6 +161,12 @@ impl IndexCheckpoint {
         Ok(())
     }
 
+    /// Clears the checkpoint of the source identified by `source_id`. Returns whether a mutation
```
Comment (fmassot) [inline quickwit-metastore/src/checkpoint.rs:164]:
```suggestion
    /// Resets the checkpoint of the source identified by `source_id`. Returns whether a mutation
```

File: quickwit-metastore/src/metastore/file_backed_metastore/file_backed_index.rs [hunk 3]
```diff
@@ -378,4 +378,9 @@ impl FileBackedIndex {
         self.metadata.delete_source(source_id)?;
         Ok(true)
     }
+
+    /// Clears the checkpoint of a source. Returns whether a mutation occurred.
```
Comment (fmassot) [inline quickwit-metastore/src/metastore/file_backed_metastore/file_backed_index.rs:382]:
```suggestion
    /// Resets the checkpoint of a source. Returns whether a mutation occurred.
```

File: quickwit-metastore/src/metastore/grpc_metastore/mod.rs [hunk 4]
```diff
@@ -406,6 +407,25 @@ impl Metastore for MetastoreGrpcClient {
             .map_err(|tonic_error| parse_grpc_error(&tonic_error))?;
         Ok(())
     }
+
+    /// Clears a source checkpoint.
```
Comment (fmassot) [inline quickwit-metastore/src/metastore/grpc_metastore/mod.rs:411]:
```suggestion
    /// Resets a source checkpoint.
```

File: quickwit-metastore/src/metastore/mod.rs [hunk 5]
```diff
@@ -179,6 +179,10 @@ pub trait Metastore: Send + Sync + 'static {
     /// If the checkpoint is missing, this does not trigger an error.
     async fn delete_source(&self, index_id: &str, source_id: &str) -> MetastoreResult<()>;
 
+    /// Clears the checkpoint of a source identified by `index_id` and `source_id`.
```
Comment (fmassot) [inline quickwit-metastore/src/metastore/mod.rs:182]:
```suggestion
    /// Resets the checkpoint of a source identified by `index_id` and `source_id`.
```

File: quickwit-proto/proto/metastore_api.proto [hunk 6]
```diff
@@ -58,6 +58,9 @@ service MetastoreApiService {
 
   // Remove source.
   rpc delete_source(DeleteSourceRequest) returns (SourceResponse);
+
+  // Clear source checkpoint.
```
Comment (fmassot) [inline quickwit-proto/proto/metastore_api.proto:62]:
```suggestion
  // Resets source checkpoint.
```

File: quickwit-proto/src/quickwit_metastore_api.rs [hunk 7]
```diff
@@ -438,6 +446,26 @@ pub mod metastore_api_service_client {
             );
             self.inner.unary(request.into_request(), path, codec).await
         }
+        /// Clear source checkpoint.
```
Comment (fmassot) [inline quickwit-proto/src/quickwit_metastore_api.rs:449]:
```suggestion
        /// Resets source checkpoint.
```

File: quickwit-proto/src/quickwit_metastore_api.rs [hunk 8]
```diff
@@ -507,6 +535,11 @@ pub mod metastore_api_service_server {
             &self,
             request: tonic::Request<super::DeleteSourceRequest>,
         ) -> Result<tonic::Response<super::SourceResponse>, tonic::Status>;
+        /// Clear source checkpoint.
```
Comment (fmassot) [inline quickwit-proto/src/quickwit_metastore_api.rs:538]:
```suggestion
        /// Resets source checkpoint.
```

**Review by fmassot — APPROVED** [review]
LGTM!

**Pull request merged**